src/snowflake/snowpark/_internal/debug_utils.py

```diff
@@ -0,0 +1,203 @@
+#
+# Copyright (c) 2012-2025 Snowflake Computing Inc. All rights reserved.
+#
+
+from functools import cached_property
+import os
+import sys
+from typing import Dict, List, Optional
+
+from snowflake.snowpark._internal.ast.batch import get_dependent_bind_ids
+from snowflake.snowpark._internal.ast.utils import __STRING_INTERNING_MAP__
+import snowflake.snowpark._internal.proto.generated.ast_pb2 as proto
+
+UNKNOWN_FILE = "__UNKNOWN_FILE__"
+SNOWPARK_PYTHON_DATAFRAME_TRANSFORM_TRACE_LENGTH = (
+    "SNOWPARK_PYTHON_DATAFRAME_TRANSFORM_TRACE_LENGTH"
+)
+
+
+class DataFrameTraceNode:
+    """A node representing a dataframe operation in the DAG that represents the lineage of a DataFrame."""
+
+    def __init__(self, batch_id: int, stmt_cache: Dict[int, proto.Stmt]) -> None:
+        self.batch_id = batch_id
+        self.stmt_cache = stmt_cache
+
+    @cached_property
+    def children(self) -> set[int]:
+        """Returns the batch_ids of the children of this node."""
+        return get_dependent_bind_ids(self.stmt_cache[self.batch_id])
+
+    def get_src(self):
+        """The source Stmt of the DataFrame described by the batch_id."""
+        stmt = self.stmt_cache[self.batch_id]
+        api_call = stmt.bind.expr.WhichOneof("variant")
+        return (
+            getattr(stmt.bind.expr, api_call).src
+            if api_call and getattr(stmt.bind.expr, api_call).HasField("src")
+            else None
+        )
+
+    def _read_file(
+        self, filename, start_line, end_line, start_column, end_column
+    ) -> str:
+        """Read the relevant code snippets of where the DataFrame was created. The filename given here
+        must have read permissions for the executing user."""
+        with open(filename) as f:
+            code_lines = []
+            for i, line in enumerate(f, 1):
+                if sys.version_info >= (3, 11):
+                    if start_line <= i <= end_line:
+                        code_lines.append(line)
+                    elif i > end_line:
+                        break
+                else:
+                    # For python 3.9/3.10, we do not extract the end line from the source code
+                    # so we just read the start line and return.
+                    if i == start_line:
+                        code_lines.append(line)
+                        break
+
+            if sys.version_info >= (3, 11):
+                if start_line == end_line:
+                    code_lines[0] = code_lines[0][start_column:end_column]
+                else:
+                    code_lines[0] = code_lines[0][start_column:]
+                    code_lines[-1] = code_lines[-1][:end_column]
+
+            code_lines = [line.rstrip() for line in code_lines]
+            return "\n".join(code_lines)
+
+    def get_source_id(self) -> str:
```

Comment:
I feel this can also be cached property, or could get_src return dynamic result?

Reply:
It can be cached_property as well. let me update that